### PR TITLE
Report undeclared lifetimes during late resolution.

### PR DESCRIPTION
**Pull request by cjgillot**
First step in https://github.com/rust-lang/rust/pull/91557

We reuse the rib design of the current resolution framework. Specific `LifetimeRib` and `LifetimeRibKind` types are introduced. The most important variant is `LifetimeRibKind::Generics`, which happens each time we encounter something which may introduce generic lifetime parameters. It can be an item or a `for<...>` binder. The `LifetimeBinderKind` specifies how this rib behaves with respect to in-band lifetimes.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -1793,83 +1794,118 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+        lifetime_ref: &ast::Lifetime,
+        outer_lifetime_ref: Option<Ident>,
+    ) {
+        debug_assert_ne!(lifetime_ref.ident.name, kw::UnderscoreLifetime);
+        let mut err = if let Some(outer) = outer_lifetime_ref {
+            let mut err = struct_span_err!(
+                self.r.session,
+                lifetime_ref.ident.span,
+                E0401,
+                "can't use generic parameters from outer function",
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:1809]:
```suggestion
                "can't use generic parameters from outer item",
```
(And in other messages below.)

I suspect that it's not always a function, impl items may also be a common case.

**Comment by petrochenkov** [comment]
r=me after addressing https://github.com/rust-lang/rust/pull/95779#discussion_r850348325 and squashing review commits into main commits.

**Comment by petrochenkov** [comment]
With this PR compiler does some double work when resolving lifetimes, and the performance should be recovered when other parts of https://github.com/rust-lang/rust/pull/91557 land.
@bors r+

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -95,7 +95,7 @@ pub enum LifetimeName {
     /// User wrote nothing (e.g., the lifetime in `&u32`).
     ///
     /// The bool indicates whether the user should have written something.
```
Comment (jackh726) [inline compiler/rustc_hir/src/hir.rs:97]: This comment is now out of date.

**Comment by pnkfelix** [comment]
Visiting for weekly performance triage

* These performance regressions were anticipated (via measurement) during review of the PR and effectively already triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**